app/src/main/java/org/lineageos/glimpse/ViewActivity.kt

```diff
@@ -49,56 +52,178 @@ class ViewActivity : AppCompatActivity() {
-                        Toast.LENGTH_SHORT
-                    ).show()
-                    finish()
-                }
+            return
+        }
 
-                return@launch
+        val dataType = intent.type ?: getContentType(uri) ?: run {
+            runOnUiThread {
+                Toast.makeText(
+                    this@ViewActivity,
+                    R.string.intent_media_type_not_found,
+                    Toast.LENGTH_SHORT
+                ).show()
+                finish()
             }
 
-            val uriType = MediaType.fromMimeType(dataType) ?: run {
-                runOnUiThread {
-                    Toast.makeText(
-                        this@ViewActivity,
-                        R.string.intent_media_type_not_supported,
-                        Toast.LENGTH_SHORT
-                    ).show()
-                    finish()
-                }
+            return
+        }
 
-                return@launch
+        val uriType = MediaType.fromMimeType(dataType) ?: run {
+            runOnUiThread {
+                Toast.makeText(
+                    this@ViewActivity,
+                    R.string.intent_media_type_not_supported,
+                    Toast.LENGTH_SHORT
+                ).show()
+                finish()
             }
 
+            return
+        }
+
+        runOnUiThread {
+            supportFragmentManager
+                .beginTransaction()
+                .replace(
+                    R.id.navHostFragment, MediaViewerFragment.newInstance(
+                        null, null, MediaUri(uri, uriType, dataType)
+                    )
+                )
+                .commit()
+        }
+    }
+
+    /**
+     * Handle a [MediaStore.ACTION_REVIEW] / [MediaStore.ACTION_REVIEW_SECURE] intent
+     * (view a media together with medias from the same bucket ID).
+     * If uri parsing from [MediaStore] fails, fallback to [handleView].
+     * Must be executed on [ioScope].
+     * @param intent The received intent
+     */
+    private fun handleReview(intent: Intent) {
+        intent.data?.let { getMediaStoreMedia(it) }?.also {
             runOnUiThread {
                 supportFragmentManager
                     .beginTransaction()
                     .replace(
                         R.id.navHostFragment, MediaViewerFragment.newInstance(
-                            null, null, MediaUri(uri, uriType, dataType)
+                            it, it.bucketId, null
                         )
                     )
                     .commit()
             }
-        }
+        } ?: handleView(intent)
     }
 
+    /**
+     * Given a [MediaStore] [Uri], parse its information and get a [Media] object.
+     * Must be executed on [ioScope].
+     * @param uri The [MediaStore] [Uri]
+     */
+    private fun getMediaStoreMedia(uri: Uri) = runCatching {
+        contentResolver.query(
+            uri,
+            arrayOf(
```

Comment:
I've been there before and was confused.
For external content you use MediaStore.MediaColumns
For MediaStore you use MediaStore.Files.FileColumns ( that extends MediaColumns )
MediaColumns doesn't define MEDIA_TYPE